components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java

```diff
@@ -394,17 +395,66 @@ private void updateFeaturesFromJSON(XWikiDocument doc, BaseObject data, XWikiCon
             // as in constructor: make unmodifiable
             this.features = Collections.unmodifiableSet(this.features);
 
-            // update the values in the document (overwriting the old list, if any)
+            // update the values in the document
+            for (String type : PHENOTYPE_PROPERTIES) {
+                data.set(type, null, context);
+            }
             for (String type : featuresMap.keySet()) {
                 data.set(type, featuresMap.get(type), context);
             }
+
+            // update features' metadata objects in document
+            updateMetaData(doc, context);
+
+            // update features' categories objects in document
+            updateCategories(doc, context);
+
             context.getWiki().saveDocument(doc, "Updated features from JSON", true, context);
 
         } catch (Exception ex) {
             this.logger.warn("Failed to update patient features from JSON [{}]: {}", ex.getMessage(), ex);
         }
     }
 
+    private void updateMetaData(XWikiDocument doc, XWikiContext context) throws XWikiException
+    {
+        doc.removeXObjects(FeatureMetadatum.CLASS_REFERENCE);
+        for (Feature feature : this.features) {
+            @SuppressWarnings("unchecked")
+            Map<String, FeatureMetadatum> metadataMap = (Map<String, FeatureMetadatum>) feature.getMetadata();
+            if (metadataMap.isEmpty() && feature.getNotes().isEmpty()) {
+                continue;
+            }
+
+            BaseObject metaObject = doc.newXObject(FeatureMetadatum.CLASS_REFERENCE, context);
+            metaObject.set(PhenoTipsFeature.META_PROPERTY_NAME, ((PhenoTipsFeature) feature).getPropertyName(),
```

Comment:
I'm re-reading the entire pull request right now, but I think whatever makes PhenoTipsFeature different from Feature should just be in Feature to start with. If a PhenoTipsFeature may have "properties", a generic Feature should as well